### ISSUE TITLE
weights for maglev's hash

### ISSUE THREAD

**Issue by AirbornWdd** [facebookincubator/katran#35]
Hi:
I configure 1 VIP and 2 real servers with weight value `1` and `10`.
But the 2 real servers filled in `ch_rings` bpf map have almost the same number (The ratio is `1:1`). 
I think [this line](https://github.com/facebookincubator/katran/blame/master/katran/lib/CHHelpers.cpp#L87) in the source code is the reason. Anyone can tell me the reason why assigning `1` to the weight of the endpoint ?

**Comment by tehnerd** [comment]
Current implementation assumes that sum of all weights must be more or less the size of hashing ring (by default 65537). So in your example on of em should have weight 6k and another ~60k (to have 1:10 ratio). Here (at fb) what we do, is using logic in control plane of balancer which is working like this:
1. Calculate sum of all real. (Eg 1+10)
2. Devide hash ring size by this sum. (65537/11 = 5957)
3. Allocate weight to each by multiple it’s original weight with number from step 2 (1*5957 for real one. 10 * 5957 for real 2)

**Comment by tehnerd** [comment]
reason why we do this:
Tl;dr - historical
Details - it is because in our initial implementation when we have tried to do it in steps we saw some unfairness in load balancing. So relying on user space to do proper weight adjustment was easier. After open sourcing katran i’ve spoke with Daniel Eisenbud (one of the author of maglev hash/paper) and he pointed me out to the exact same flow/requirement of our implementation and told me how it looks like in G. Unfortunately changing of hashing would break current deployment (during rollout phase; as hashes would stop to be consistent between old and new version) and in terms of stability / fairness I have not saw any big benefits or their way vs ours (G way vs what we have here) so we decided to leave it as is. There are plans to make hashing function  pkugable one day, so we would be able to add any other hashing function/ implementation of maglevs hashing

**Comment by AirbornWdd** [comment]
I see. Thanks very much.
